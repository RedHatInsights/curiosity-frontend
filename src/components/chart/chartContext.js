import React, { useCallback, useContext } from 'react';
import { helpers } from '../../common';

/**
 * Chart context.
 *
 * @type {React.Context<{}>}
 */
const DEFAULT_CONTEXT = [
  { chartContainerRef: helpers.noop, chartSettings: {}, chartTooltipRef: helpers.noop, dataSetsToggle: [] },
  helpers.noop
];

const ChartContext = React.createContext(DEFAULT_CONTEXT);

/**
 * Get an updated chart context.
 *
 * @returns {React.Context<{}>}
 */
const useChartContext = () => useContext(ChartContext);

/**
 * ToDo: reevaluate this alternative pattern of passing hooks as options, helps testing
 */
/**
 * Track, show, and hide chart data layers.
 *
 * @param {object} hooks
 * @param {Function} hooks.useChartContext
 * @returns {{onRevert: Function, onToggle: Function, getIsToggled: Function, dataSetsToggle: object,
 *     onHide: Function}}
 */
const useToggleData = ({ useChartContext: useAliasChartContext = useChartContext } = {}) => {
  const { dataSetsToggle: contextDataSetsToggle = [] } = useAliasChartContext();
  const [dataSetsToggle, setDataSetsToggle] = contextDataSetsToggle;

  /**
   * Hide a graph layer.
   *
   * @type {(function(*): void)|*}
   */
  const onHide = useCallback(
    id => {
      setDataSetsToggle(prevState => ({ ...prevState, [id]: true }));
    },
    [setDataSetsToggle]
  );

  // ToDo: re-evaluate useCallback here.
  /**
   * Reset graph layers.
   *
   * @type {(function(): void)|*}
   */
  const onRevert = useCallback(() => {
    setDataSetsToggle(() => ({}));
  }, [setDataSetsToggle]);

  /**
   * Hide/show graph layers.
   *
   * @type {function(*): boolean}
   */
  const onToggle = useCallback(
    id => {
      const updatedToggle = !dataSetsToggle?.[id];
      setDataSetsToggle(prevState => ({ ...prevState, [id]: updatedToggle }));
      return updatedToggle;
    },
    [dataSetsToggle, setDataSetsToggle]
  );

  // ToDo: review return undefined if doesn't exist
  /**
   * Graph layer status.
   *
   * @type {function(*): boolean}
   */
  const getIsToggled = useCallback(id => dataSetsToggle?.[id] || false, [dataSetsToggle]);

  return {
    ...{ dataSetsToggle },
    onHide,
    onRevert,
    onToggle,
    getIsToggled
  };
};

const context = {
  ChartContext,
  DEFAULT_CONTEXT,
  useChartContext,
  useToggleData
};

export { context as default, context, ChartContext, DEFAULT_CONTEXT, useChartContext, useToggleData };
