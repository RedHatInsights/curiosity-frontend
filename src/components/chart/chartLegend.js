import React from 'react';
import _cloneDeep from 'lodash/cloneDeep';
import { useChartContext, useToggleData } from './chartContext';

/**
 * Wrapper for rendering an HTML based legend.
 *
 * @returns {Node}
 */
const ChartLegend = () => {
  const { getIsToggled, onHide, onRevert, onToggle } = useToggleData();
  const { chartSettings = {} } = useChartContext();
  const { chartLegend, dataSets } = chartSettings;

  if (!chartLegend) {
    return null;
  }

  const legendProps = {
    datum: { dataSets: _cloneDeep(dataSets) },
    chart: {
      hide: onHide,
      revert: onRevert,
      toggle: onToggle,
      isToggled: getIsToggled
    }
  };

  return (
    <div className="curiosity-chartarea__legend">
      {(React.isValidElement(chartLegend) && React.cloneElement(chartLegend, { ...legendProps })) ||
        chartLegend({ ...legendProps })}
    </div>
  );
};

/**
 * Prop types.
 */
ChartLegend.propTypes = {};

/**
 * Default props.
 */
ChartLegend.defaultProps = {};

export { ChartLegend as default, ChartLegend };
