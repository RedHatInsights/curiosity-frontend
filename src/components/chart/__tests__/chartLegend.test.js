import React from 'react';
import { shallow } from 'enzyme';
import { ChartLegend } from '../chartLegend';
import * as chartContext from '../chartContext';

describe('ChartLegend Component', () => {
  it('should render a basic component', () => {
    const props = {};
    const component = shallow(<ChartLegend {...props} />);

    expect(component.render()).toMatchSnapshot('basic');
  });

  it('should pass legend props', async () => {
    const props = {};

    const mockLegend = jest.fn();
    const mockContextValue = {
      chartSettings: {
        chartLegend: mockLegend,
        dataSets: [
          {
            data: [
              {
                x: 1,
                xAxisLabel: '1 x axis label',
                y: 1
              },
              {
                x: 2,
                xAxisLabel: '2 x axis label',
                y: 2
              }
            ],
            fill: '#ipsum',
            id: 'loremGraph',
            isStacked: true,
            stroke: '#lorem',
            strokeWidth: 2
          },
          {
            data: [
              {
                x: 1,
                xAxisLabel: '1 x axis label',
                y: 1
              },
              {
                x: 2,
                xAxisLabel: '2 x axis label',
                y: 2
              }
            ],
            fill: '#ipsum',
            id: 'ipsumGraph',
            isThreshold: true,
            stroke: '#lorem',
            strokeDasharray: '4,3',
            strokeWidth: 3
          }
        ]
      }
    };

    const mock = jest.spyOn(chartContext, 'useChartContext').mockImplementation(() => mockContextValue);

    await shallowHookComponent(<ChartLegend {...props} />);
    expect(mockLegend.mock.calls).toMatchSnapshot('passed props');

    mock.mockClear();
  });

  it('should handle legend elements', async () => {
    const props = {};

    const mockLegend = <div>lorem ipsum</div>;
    const mockContextValue = {
      chartSettings: {
        chartLegend: mockLegend,
        dataSets: [
          {
            data: [
              {
                x: 1,
                xAxisLabel: '1 x axis label',
                y: 1
              },
              {
                x: 2,
                xAxisLabel: '2 x axis label',
                y: 2
              }
            ],
            id: 'loremGraph'
          }
        ]
      }
    };

    const mock = jest.spyOn(chartContext, 'useChartContext').mockImplementation(() => mockContextValue);

    const component = await shallowHookComponent(<ChartLegend {...props} />);
    expect(component).toMatchSnapshot('element');

    mock.mockClear();
  });
});
