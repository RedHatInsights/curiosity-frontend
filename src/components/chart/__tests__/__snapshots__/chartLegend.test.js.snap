// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartLegend Component should handle legend elements: element 1`] = `
<div
  className="curiosity-chartarea__legend"
>
  <div
    chart={
      Object {
        "hide": [Function],
        "isToggled": [Function],
        "revert": [Function],
        "toggle": [Function],
      }
    }
    datum={
      Object {
        "dataSets": Array [
          Object {
            "data": Array [
              Object {
                "x": 1,
                "xAxisLabel": "1 x axis label",
                "y": 1,
              },
              Object {
                "x": 2,
                "xAxisLabel": "2 x axis label",
                "y": 2,
              },
            ],
            "id": "loremGraph",
          },
        ],
      }
    }
  >
    lorem ipsum
  </div>
</div>
`;

exports[`ChartLegend Component should pass legend props: passed props 1`] = `
Array [
  Array [
    Object {
      "chart": Object {
        "hide": [Function],
        "isToggled": [Function],
        "revert": [Function],
        "toggle": [Function],
      },
      "datum": Object {
        "dataSets": Array [
          Object {
            "data": Array [
              Object {
                "x": 1,
                "xAxisLabel": "1 x axis label",
                "y": 1,
              },
              Object {
                "x": 2,
                "xAxisLabel": "2 x axis label",
                "y": 2,
              },
            ],
            "fill": "#ipsum",
            "id": "loremGraph",
            "isStacked": true,
            "stroke": "#lorem",
            "strokeWidth": 2,
          },
          Object {
            "data": Array [
              Object {
                "x": 1,
                "xAxisLabel": "1 x axis label",
                "y": 1,
              },
              Object {
                "x": 2,
                "xAxisLabel": "2 x axis label",
                "y": 2,
              },
            ],
            "fill": "#ipsum",
            "id": "ipsumGraph",
            "isThreshold": true,
            "stroke": "#lorem",
            "strokeDasharray": "4,3",
            "strokeWidth": 3,
          },
        ],
      },
    },
  ],
]
`;

exports[`ChartLegend Component should render a basic component: basic 1`] = `null`;
